### PR TITLE
[receiver/mongodbreceiver]: Add Lock Metrics

### PR DESCRIPTION
**Pull request by sohampachpande**
#### Features Added: 
Lock metrics allow users to monitor how long operations wait for acquiring locks and if the database system is affected by queries acquiring locks. This PR adds the following metrics to the MongoDB receiver - 

| Name | Description | Unit | Type | Attributes |
| ---- | ----------- | ---- | ---- | ---------- |
| **mongodb.locks.acquire_count** | Number of times the lock was acquired in the specified mode. | {count} | Sum(Int) | <ul> <li>database</li> <li>lock_type</li> <li>lock_mode</li> </ul> |
| **mongodb.locks.acquire_wait_count** | Number of times the lock acquisitions encountered waits because the locks were held in a conflicting mode. | {count} | Sum(Int) | <ul> <li>database</li> <li>lock_type</li> <li>lock_mode</li> </ul> |
| **mongodb.locks.deadlock_count** | Number of times the lock acquisitions encountered deadlocks. | {count} | Sum(Int) | <ul> <li>database</li> <li>lock_type</li> <li>lock_mode</li> </ul> |
| **mongodb.locks.time_acquiring_micros** | Cumulative wait time in microseconds for the lock acquisitions. | ms | Sum(Int) | <ul> <li>database</li> <li>lock_type</li> <li>lock_mode</li> </ul> |

Please note that the lock metrics have multiple attributes and mongodb exposes the metric value only when the metric value is non zero. Accordingly, the scraper raises a "Key not found" error whenever any of the above metrics (along with attributes) is not exposed by mongoDB. Accordingly, the locks metrics are only exported if a nonzero metric value is exposed by MongoDB diagnostic commands. The test cases reflect this behaviour.

#### Compatibility 

- The Lock metrics are compatible with MongoDB versions - 3+, 4.0+, 5.0 and 6.0. 
- The lock types - parallel_batch_write_mode and replication_state_transition are only available with MongoDB v4.2+

#### Testing:

The integration was tested locally with MongoDB v6.0.0.

Test cases for scraper are updated. Two key changes are: 
1. Fixing a bug in comparing partial errors - the first error in both actual and expected error had a missing leading whitespace due to choice of delimiter. Updated delimiter to ensure all errors follow same format. 
2. MongoDB does not exclude lock related metrics if the metric value is 0 in its serverStatus diagnostic calls. Accordingly, the successful scrap test case checks for partial error related to absent lock metrics. 

#### Documentation:
The details of the Lock metrics collected are generated using mdatagen and added to the documentation.

### CODE REVIEW

File: receiver/mongodbreceiver/metadata.yaml
Comment (djaglowski) [inline receiver/mongodbreceiver/metadata.yaml:41]: This does not communicate anything more than the attribute name. Please provide a more detailed description.

File: receiver/mongodbreceiver/metadata.yaml
Comment (djaglowski) [inline receiver/mongodbreceiver/metadata.yaml:52]: This does not communicate anything more than the attribute name. Please provide a more detailed description.

File: receiver/mongodbreceiver/metadata.yaml
Comment (djaglowski) [inline receiver/mongodbreceiver/metadata.yaml:294]:
```suggestion
    attributes: [database, lock_type, lock_mode]
    
```

File: receiver/mongodbreceiver/metadata.yaml
Comment (djaglowski) [inline receiver/mongodbreceiver/metadata.yaml:279]: Microseconds or ms?

Either way, typically do not include the units in the descriptions, since there is a dedicated field for it.

File: receiver/mongodbreceiver/metadata.yaml
Comment (djaglowski) [inline receiver/mongodbreceiver/metadata.yaml:266]: Based on the description, I think this is likely monotonic. Does it ever decrease (aside from a rollover or reset)?

File: receiver/mongodbreceiver/metadata.yaml
Comment (djaglowski) [inline receiver/mongodbreceiver/metadata.yaml:275]: Same, likely monotonic

File: receiver/mongodbreceiver/metadata.yaml
Comment (djaglowski) [inline receiver/mongodbreceiver/metadata.yaml:284]: Same, likely monotonic

File: receiver/mongodbreceiver/metadata.yaml
Comment (djaglowski) [inline receiver/mongodbreceiver/metadata.yaml:293]: Same, likely monotonic

File: receiver/mongodbreceiver/metadata.yaml
Comment (djaglowski) [inline receiver/mongodbreceiver/metadata.yaml:259]:
```suggestion
  mongodb.lock.acquire.count:
```

File: receiver/mongodbreceiver/metadata.yaml
Comment (djaglowski) [inline receiver/mongodbreceiver/metadata.yaml:268]:
```suggestion
  mongodb.lock.acquire.wait_count:
```

File: receiver/mongodbreceiver/metadata.yaml
Comment (djaglowski) [inline receiver/mongodbreceiver/metadata.yaml:277]:
```suggestion
  mongodb.lock.acquire.time:
```

File: receiver/mongodbreceiver/metadata.yaml
Comment (djaglowski) [inline receiver/mongodbreceiver/metadata.yaml:286]:
```suggestion
  mongodb.lock.deadlock.count:
```

File: receiver/mongodbreceiver/scraper_test.go
Comment (djaglowski) [inline receiver/mongodbreceiver/scraper_test.go:197]: Why is this changed?

File: receiver/mongodbreceiver/scraper_test.go
Comment (djaglowski) [inline receiver/mongodbreceiver/scraper_test.go:232]: Why can't we test these metrics?

File: receiver/mongodbreceiver/scraper_test.go
Comment (djaglowski) [inline receiver/mongodbreceiver/scraper_test.go:250]: These changes appear unrelated.

File: receiver/mongodbreceiver/scraper_test.go
Comment (sohampachpande) [reply receiver/mongodbreceiver/scraper_test.go:232]: We do test these metrics. Some of the lock metrics added in this PR are not necessarily published at all times. 

MongoDB does not exclude lock related metrics if the metric value is 0 in its serverStatus diagnostic calls. Accordingly, the successful scrap test case would have partial errors and the test case checks for partial error related to absent lock metrics.

File: receiver/mongodbreceiver/scraper_test.go
Comment (sohampachpande) [reply receiver/mongodbreceiver/scraper_test.go:250]: It is unrelated but necessary. 

The first error string in both actual and expected error has a missing leading whitespace due to choice of delimiter. This change has updated delimiter to ensure all errors follow same format.

should this be a different PR?

File: receiver/mongodbreceiver/metadata.yaml
Comment (sohampachpande) [reply receiver/mongodbreceiver/metadata.yaml]: Updated description with more details

File: receiver/mongodbreceiver/metadata.yaml
Comment (sohampachpande) [reply receiver/mongodbreceiver/metadata.yaml]: Updated description with more details

File: receiver/mongodbreceiver/metadata.yaml
Comment (sohampachpande) [reply receiver/mongodbreceiver/metadata.yaml]: Got it, its microseconds. Updated description accordingly and set unit as "microseconds"

File: receiver/mongodbreceiver/metadata.yaml
Comment (sohampachpande) [reply receiver/mongodbreceiver/metadata.yaml]: you are right, updated the code.

File: receiver/mongodbreceiver/metadata.yaml
Comment (sohampachpande) [reply receiver/mongodbreceiver/metadata.yaml]: Updated

File: receiver/mongodbreceiver/metadata.yaml
Comment (sohampachpande) [reply receiver/mongodbreceiver/metadata.yaml]: Updated

File: receiver/mongodbreceiver/metadata.yaml
Comment (sohampachpande) [reply receiver/mongodbreceiver/metadata.yaml]: Updated

File: receiver/mongodbreceiver/scraper_test.go
Comment (sohampachpande) [reply receiver/mongodbreceiver/scraper_test.go:232]: Do you have any thoughts on how to deal with such metrics? As we add new metrics to mongodb receiver we will have to deal with metrics which are not always published

File: receiver/mongodbreceiver/metadata.yaml
Comment (sohampachpande) [reply receiver/mongodbreceiver/metadata.yaml]: Updated the metric name as per suggestion

File: receiver/mongodbreceiver/metadata.yaml
Comment (sohampachpande) [reply receiver/mongodbreceiver/metadata.yaml]: Updated the metric name as per suggestion

File: receiver/mongodbreceiver/metadata.yaml
Comment (sohampachpande) [reply receiver/mongodbreceiver/metadata.yaml]: Updated the metric name as per suggestion

File: receiver/mongodbreceiver/metadata.yaml
Comment (sohampachpande) [reply receiver/mongodbreceiver/metadata.yaml]: Updated the metric name as per suggestion

File: receiver/mongodbreceiver/scraper_test.go
Comment (djaglowski) [reply receiver/mongodbreceiver/scraper_test.go:232]: I think it is not an error then if we expect it will not be present in some circumstances.

**Comment by djaglowski** [comment]
@sohampachpande, do you plan to continue working on this PR? These would be valuable metrics.

File: receiver/mongodbreceiver/scraper_test.go
Comment (sohampachpande) [reply receiver/mongodbreceiver/scraper_test.go]: Agreed, modifying code to not raise a error when metrics which are not present and are not expected to be always published

Comment (sohampachpande) [reply receiver/mongodbreceiver/scraper_test.go]: Agreed, I am updating the code to not raise an error when a metric key is absent and the metric is not expected to be present at all times.

**Comment by sohampachpande** [comment]
Hi! I am planning to complete this PR and made some changes today. 

I accidently force pushed some of my changes. There was a conflict with _/receiver/mongodbreceiver/internal/metadata/generated_metrics.go_ and I decided to use the file on remote. How should I proceed now? Do I need to create a new PR?

**Comment by davidji99** [comment]
@sohampachpande Happy to continue this PR for you if you'd me to do so.

**Comment by sohampachpande** [comment]
@davidji99 Feel free to contribute to this PR. I am waiting for code owner review and workflows to execute!

File: receiver/mongodbreceiver/scraper_test.go
Comment (djaglowski) [inline receiver/mongodbreceiver/scraper_test.go:118]: Let's please move these back here since it appears this change was unnecessary for the pr.

File: receiver/mongodbreceiver/metrics.go
Comment (djaglowski) [inline receiver/mongodbreceiver/metrics.go:340]: nit: Let's have the var name correlate to the version. Same issue is present on the other new `record` funcs.
```suggestion
	mongo32, _ := version.NewVersion("3.2")
	if s.mongoVersion.LessThan(mongo32) {
```

File: receiver/mongodbreceiver/metadata.yaml
Comment (djaglowski) [reply receiver/mongodbreceiver/metadata.yaml:294]: Just need a newline at the end of the file.

**Review by djaglowski — COMMENTED** [review]
I missed that you had pushed again. 

This looks good except for a few nits & minor checks failing.

File: receiver/mongodbreceiver/metrics.go
Comment (djaglowski) [inline receiver/mongodbreceiver/metrics.go:369]:
```suggestion
	mongo32, _ := version.NewVersion("3.2")
```

File: receiver/mongodbreceiver/metrics.go
Comment (djaglowski) [inline receiver/mongodbreceiver/metrics.go:399]:
```suggestion
	mongo32, _ := version.NewVersion("3.2")
```

File: receiver/mongodbreceiver/metrics.go
Comment (djaglowski) [inline receiver/mongodbreceiver/metrics.go:429]:
```suggestion
	mongo32, _ := version.NewVersion("3.2")
```

**Comment by davidji99** [comment]
@djaglowski 

I'm helping the original PR author get this merged.

Question: I didn't happen to see a helper method for my issue but is there a suggestion to address a scenario where the `integration_test` for the mongodb receiver 4.0 returns different lengths of data points? For example, this are the test outputs:

```
        	Error:      	Received unexpected error:
        	            	the following errors occurred:
        	            	 -  datapoints for metric: `mongodb.lock.acquire.wait_count`, do not match expected
        	            	 -  number of datapoints does not match expected: 3, actual: 1
```

```
        	Error:      	Received unexpected error:
        	            	the following errors occurred:
        	            	 -  datapoints for metric: `mongodb.lock.acquire.wait_count`, do not match expected
        	            	 -  number of datapoints does not match expected: 3, actual: 2
```

**Comment by djaglowski** [comment]
> is there a suggestion to address a scenario where the integration_test for the mongodb receiver 4.0 returns different lengths of data points?

Unfortunately I don't think there is. 

In the vast majority of cases, this would indicate an actual problem, so we should start by making sure we understand why it is that we are getting a variable number of data points for this metric.

If this is expected behavior, it is likely due to these data points being created only when wait event actually happen. For example, during one run of the integration test, one database may experience a wait event. On another run, three databases experience a wait event. 

In any case, let's pin down why this is happening and make sure it really is a special case we need to handle, and not just an indication of a real problem that needs to be addressed.

**Comment by davidji99** [comment]
> For example, during one run of the integration test, one database may experience a wait event. On another run, three databases experience a wait event.

This scenario you described here could very well be emulated in real-life. Due any number of factors, the wait event could or could not be present for a given database.

Also, what's odd is this inconsistent behavior does not seem to be present when I added a new testcase using mongo4.2. Maybe, the behavior is different between minor `v4` versions.

**Comment by djaglowski** [comment]
> this inconsistent behavior does not seem to be present when I added a new testcase using mongo4.2

We'd be sidestepping the problem, but I think it'd be reasonable to validate the metric against this version only, for now.

In order to handle this type of "variable number of data points" case in the future, we will need to implement a new `CompareOption` in the `scrapertest` package that indicates that a given metric may have a variable number of data points.

**Comment by davidji99** [comment]
I'm circling back to this.

File: .chloggen/add-more-mongodb-lock-metrics.yaml
Comment (djaglowski) [inline .chloggen/add-more-mongodb-lock-metrics.yaml:21]: Please resolve. This should fix the changelog check.

**Review by djaglowski — COMMENTED** [review]
LGTM overall. Just need to disable metrics by default and fix the changelog entry

File: .chloggen/add-more-mongodb-lock-metrics.yaml
Comment (davidji99) [reply .chloggen/add-more-mongodb-lock-metrics.yaml:21]: Missed this. I'll remove.

File: receiver/mongodbreceiver/metadata.yaml
Comment (davidji99) [reply receiver/mongodbreceiver/metadata.yaml:262]: @djaglowski Is the new guideline that all new metrics to an existing receiver be disabled by default instead of enabled as it used to be?

Comment (dmitryax) [reply receiver/mongodbreceiver/metadata.yaml:262]: @davidji99 yes, that the guidelines for beta and stable components. If you think that the new metrics essential to be enabled by default, that should be applied in two steps. See https://github.com/open-telemetry/opentelemetry-collector/pull/6358/files#diff-32d5b2373095d70f3e09a70a62829841135a4b9c9f19ddec516679d88e0e53c3R121 for more details

**Comment by davidji99** [comment]
@djaglowski this is ready for review again.

File: receiver/mongodbreceiver/testdata/scraper/expected.json
Comment (djaglowski) [inline receiver/mongodbreceiver/testdata/scraper/expected.json:24]: There don't appear to be any meaningful changes in this file, so please revert them.

File: receiver/mongodbreceiver/testdata/integration/expected.5_0.json
Comment (djaglowski) [inline receiver/mongodbreceiver/testdata/integration/expected.5_0.json:1914]:
```suggestion
}

```

File: receiver/mongodbreceiver/testdata/integration/expected.4_0.json
Comment (djaglowski) [inline receiver/mongodbreceiver/testdata/integration/expected.4_0.json:2006]:
```suggestion
}

```

**Review by djaglowski — COMMENTED** [review]
LGTM except a couple extraneous changes in the expected json files.

**Comment by davidji99** [comment]
@djaglowski Done. Ready for review.

**Comment by davidji99** [comment]
@djaglowski Friendly reminder. Thanks!

**Review by djaglowski — APPROVED** [review]

**Pull request merged**
